### ISSUE TITLE
Dapper Calls via Execute Aren't Instrumented

### ISSUE THREAD

**Issue by bretkoppel** [DataDog/dd-trace-dotnet#661]
**Describe the bug**
Database calls using the [Dapper](https://github.com/StackExchange/Dapper) library's [`.Execute...` method](https://github.com/StackExchange/Dapper/blob/1.60.1/Dapper/SqlMapper.cs#L440) seem to delegate to [`IDbCommand.ExecuteNonQuery()`]((https://github.com/StackExchange/Dapper/blob/1.60.1/Dapper/SqlMapper.cs#L2827)), but they don't appear when traced. Calls via `.Query...` _are_ instrumented. This happens against both Sql Server and Postgres(using npgsql).

**To Reproduce**
Super basic example gist is [here](https://gist.github.com/bretkoppel/932117202edbc1167d3312d638734b6a). You'll need to do some connection string swapping but hopefully this is clear enough.

**Expected behavior**
4 spans are created(2 for `.Execute`; 2 for `.Query`)

**Screenshots**
https://cl.ly/592416576d1c

**Runtime environment (please complete the following information):**
- Instrumentation mode: Automatic
- Tracer version: 1.13.3
- OS: Windows Server 2012 R2
- CLR: .NET Core 2.2

**Additional context**
This is a follow-up to https://github.com/DataDog/dd-trace-dotnet/issues/581 and https://github.com/DataDog/dd-trace-dotnet/issues/540

@bobuva Pinging you as you asked! Please let me know if I can add anything.

**Comment by bobuva** [comment]
@bretkoppel  Thanks very much! We're planning on releasing the other Postgres / Dapper support that is currently in a prerelease, soon. But we've got this issue on our backlog to investigate. Is it currently blocking you?

**Comment by bretkoppel** [comment]
@bobuva Aside from the visibility issue, this isn't currently blocking us. We are hoping to ramp up our deployment of the tracer pretty rapidly in the coming months and the internal consumer audience will be much larger at that point, so it would be a little hairier.

**Comment by bobuva** [comment]
Ok, good to know. I'll let you know when we have the non-query stuff working.

**Comment by bretkoppel** [comment]
@bobuva Just wanted to follow up because I had a chance to play with this a bit more yesterday. I ended up adding a straight-to-npgsql ExecuteNonQuery in addition to my dapper-based ones:

```
using (var cmd = new NpgsqlCommand("select 1,1,1", connection))
{
    if (connection.State != System.Data.ConnectionState.Open)
        connection.Open();
    Console.WriteLine(cmd.ExecuteNonQuery());
}
```

This didn't get instrumented. I did pull some additional log data:

```
[TraceTesting.exe] 14004: [debug] JITCompilationStarted: function_id=140703505992608 token=100663297 name=TraceTesting.Program.Main()
[TraceTesting.exe] 14004: [info] JITCompilationStarted skipping method: Method replacement found but the managed profiler has not yet been loaded. function_id=140703505992608 token=100663297 target_name=System.Data.Common.DbCommand.ExecuteNonQuery()
...
[TraceTesting.exe] 14004: [debug] JITCompilationStarted: function_id=140703509263328 token=100663980 name=Npgsql.NpgsqlCommand.ExecuteNonQuery()
[TraceTesting.exe] 14004: [debug] JITCompilationStarted skipping function call: argument counts don't match. function_id=140703509263328 token=100663980 target_name=Npgsql.NpgsqlCommand.ExecuteNonQuery() expected_number_args=0 target_arg_count=2
[TraceTesting.exe] 14004: [debug] JITCompilationStarted: function_id=140703509263344 token=100663982 name=Npgsql.NpgsqlCommand.ExecuteNonQuery()
[TraceTesting.exe] 14004: [debug] JITCompilationStarted: function_id=140703519467088 token=100666370 name=<ExecuteNonQuery>d__92.MoveNext()
[TraceTesting.exe] 14004: [debug] JITCompilationStarted skipping function call: argument counts don't match. function_id=140703519467088 token=100666370 target_name=Npgsql.NpgsqlCommand.ExecuteReaderAsync() expected_number_args=1 target_arg_count=3
[TraceTesting.exe] 14004: [debug] JITCompilationStarted skipping function call: argument counts don't match. function_id=140703519467088 token=100666370 target_name=Npgsql.NpgsqlCommand.ExecuteReaderAsync() expected_number_args=2 target_arg_count=3
```

Hope this helps!

FWIW, I don't know enough about the profiling API to understand how it operates with inlining, but was wondering if maybe the `AggressiveInlining` attribute on `ExecuteNonQuery`([ref](https://github.com/npgsql/npgsql/blob/afe2ac769f0a4d9b6024a6ec181ac5bbeeab29f0/src/Npgsql/NpgsqlCommand.cs#L985-L986)) would cause the argument mismatch. Just a naive thought.

**Comment by lucaspimentel** [comment]
These log messages will be very useful, @bretkoppel. Thanks!

**Comment by HaydnDias** [comment]
Hi guys, is there any update or progress on this? It's currently the only thing we're missing in tracing and would be amazing is we could get this functionality working.

Many thanks!

**Comment by lucaspimentel** [comment]
Hello, everyone. We just merged a big PR (#753) that adds missing ADO.NET instrumentation for several scenarios. The changes are currently available in [1.18.4-prerelease](https://github.com/DataDog/dd-trace-dotnet/releases/tag/v1.18.4-prerelease) if you're interesting in testing them before we do a full release. Thanks for your patience with this one!

**Comment by bretkoppel** [comment]
@lucaspimentel Apologies for the delay - I finally had a chance to try out 1.19.0 and we are seeing these missing spans. Thank you! Feel free to close this out.
